daphne/src/roles.rs

```diff
@@ -198,9 +213,19 @@ pub trait DapLeader<S>: DapAuthorizedSender<S> + DapAggregator<S> {
     /// Handle HTTP POST to `/upload`. The input is the encoded report sent in the body of the HTTP
     /// request.
     async fn http_post_upload(&self, req: &DapRequest<S>) -> Result<(), DapAbort> {
+        // Check whether the DAP version in the request is not Unknown.
+        if req.version == DapVersion::Unknown {
+            return Err(DapAbort::InvalidProtocolVersion);
+        }
+
         let report = Report::get_decoded(req.payload.as_ref())?;
-        if self.get_task_config_for(&report.task_id).is_none() {
-            return Err(DapAbort::UnrecognizedTask);
+        let task_config = self
+            .get_task_config_for(&report.task_id)
+            .ok_or(DapAbort::UnrecognizedTask)?;
+
+        // Check whether the DAP version in the request matches the one in task config.
+        if task_config.version != req.version {
+            return Err(DapAbort::InvalidProtocolVersion);
         }
```

Comment:
It looks like you forgot to remove this code?

Reply:
Do we want to remove this check? I think it is important that the version in the request and the one that is agreed in the task config are the same.

Reply:
Oops, sorry I forgot to remove this. I realized after making this comment that you're checking first whether the version is known and then whether the version is supported, which is exactly right.

Now, it would be fair to delegate the latter to the function that needs it. For example, if decoding needs the version number, then we could just pass the version number there. But I think what you have is fine too.